### PR TITLE
hevm: --cov-match flag

### PR DESCRIPTION
**Pull request by TransmissionsDev**
My first attempt at some substantive haskell contributions. thanks @d-xo for all the guidance 🙏 ❤️

### CODE REVIEW

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 1]
```diff
@@ -298,6 +300,7 @@ unitTestOptions cmd testFile = do
     , EVM.UnitTest.askSmtIters = askSmtIterations cmd
     , EVM.UnitTest.smtTimeout = smttimeout cmd
     , EVM.UnitTest.solver = solver cmd
+    , EVM.UnitTest.covMatch = covMatch cmd >>= pure . pack
```
Comment (TransmissionsDev) [inline src/hevm/hevm-cli/hevm-cli.hs:303]: Is there a more compact way to write this?

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 2]
```diff
@@ -593,24 +596,32 @@ dappCoverage opts _ solcFile =
...
         mapM_ f (Map.toList (coverageReport dapp covs))
       Nothing ->
         error ("Failed to read Solidity JSON for `" ++ solcFile ++ "'")
 
+shouldPrintCoverage :: Maybe Text -> Text -> Bool
+shouldPrintCoverage (Just covMatch) name = regexMatches covMatch name
+shouldPrintCoverage Nothing name = not (isCoverageIrrelevant name)
+
+isCoverageIrrelevant :: Text -> Bool
+isCoverageIrrelevant name = Text.isPrefixOf "src/external/" name || Text.isPrefixOf "src/tests/" name || Text.isPrefixOf "src/test/" name || Text.isPrefixOf "lib/" name || Text.isSuffixOf ".t.sol" name
```
Comment (TransmissionsDev) [inline src/hevm/hevm-cli/hevm-cli.hs:623]: Is there a more haskell-y way to write this? Maybe a function that takes an array of text called `areAnyPrefixOf`?

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 1 above)
Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:303]:
```suggestion
    , EVM.UnitTest.covMatch = pack <$> covMatch cmd
```

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:303]: How about this?

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs:303]: infix fmap, smart!

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 3]
```diff
@@ -593,24 +596,38 @@ dappCoverage opts _ solcFile =
...
         mapM_ f (Map.toList (coverageReport dapp covs))
       Nothing ->
         error ("Failed to read Solidity JSON for `" ++ solcFile ++ "'")
 
+shouldPrintCoverage :: Maybe Text -> Text -> Bool
+shouldPrintCoverage (Just covMatch) file = regexMatches covMatch file
+shouldPrintCoverage Nothing file = not (isCoverageIrrelevant file)
+
+isCoverageIrrelevant :: Text -> Bool
+isCoverageIrrelevant file = isAnyPrefixOf ["src/test/", "src/tests/", "lib/"] file || Text.isSuffixOf ".t.sol" file
```
Comment (TransmissionsDev) [inline src/hevm/hevm-cli/hevm-cli.hs:623]: Is there a better way to do this? repeating `file` feels dirty hehe

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 4]
```diff
@@ -593,24 +596,38 @@ dappCoverage opts _ solcFile =
...
+
+isAnyPrefixOf :: [Text] -> Text -> Bool
+isAnyPrefixOf [] _ = False
+isAnyPrefixOf (x:xs) t
+  | Text.isPrefixOf x t = True
+  | otherwise = isAnyPrefixOf xs t
```
Comment (TransmissionsDev) [inline src/hevm/hevm-cli/hevm-cli.hs:629]: better way than guards to do this? 

or could this whole function be refactored into some sort of fold or map operation?

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 3 above)
Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:623]: Doesnt look too bad

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 4 above)
Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:629]: Yeah, you could do:
```suggestion
isAnyPrefixOf prefixes t = any (flip Text.isPrefixOf t) prefixes
```

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:629]: Or smth like this

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs:629]: oo dope

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 5]
```diff
@@ -593,24 +596,36 @@ dappCoverage opts _ solcFile =
...
 
+shouldPrintCoverage :: Maybe Text -> Text -> Bool
+shouldPrintCoverage (Just covMatch) file = regexMatches covMatch file
+shouldPrintCoverage Nothing file = not (isTestOrLib file)
+
+isTestOrLib :: Text -> Bool
+isTestOrLib file = isAnyPrefixOf ["src/test/", "src/tests/", "lib/"] file || Text.isSuffixOf ".t.sol" file
+
+isAnyPrefixOf :: [Text] -> Text -> Bool
+isAnyPrefixOf [] _ = False
```
Comment (TransmissionsDev) [inline src/hevm/hevm-cli/hevm-cli.hs:626]: We don't need this anymore do we?

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs:626]: compiled locally without, removing, lmk if i should revert ;p

File: src/hevm/hevm-cli/hevm-cli.hs [hunk 6]
```diff
@@ -593,24 +596,36 @@ dappCoverage opts _ solcFile =
...
-                Nothing -> putStr "\x1b[38;5;240m" -- Gray (Coverage status isn't relevant)
-                Just _ ->
-                  case n of
-                    -1 -> putStr "\x1b[38;5;240m" -- Gray (Coverage status isn't relevant)
-                    0  -> putStr "\x1b[31m" -- Red (Uncovered)
-                    _  -> putStr "\x1b[32m" -- Green (Covered)
-              Char8.putStrLn bs
-            putStrLn ""
-
+            if shouldPrintCoverage (EVM.UnitTest.covMatch opts) k then do
```
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs:599]: when shouldPrintCoverage
could also be cute here

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 5 above)
Comment (MrChico) [inline src/hevm/hevm-cli/hevm-cli.hs]: You dont need this case anymore btw

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs]: yup asked up above hehe, removed :)

File: src/hevm/hevm-cli/hevm-cli.hs (hunk 6 above)
Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: don't follow 😓

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: Yup, `when x m = if x then m else pure ()`

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: ohhh neat lemme give it a shot

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: also, in general, hoogle is your friend. The nix shell also contains language server which I think may also have hoogle built in and you can get it from whatever editor youre using

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: hmm ok how does this work when m is a multiline block like we have atm?

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: Like this:
```
when x $
   do m
       ...
```

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs:599]: And when you are done you just go back to previous indentation level

Comment (TransmissionsDev) [reply src/hevm/hevm-cli/hevm-cli.hs]: aight just gave it a shot, in this case putting `do` on the same line as the `when` felt more natural, but lmk if you'd prefer on it a newline 😅

Comment (MrChico) [reply src/hevm/hevm-cli/hevm-cli.hs]: sounds good too!

**Comment by TransmissionsDev** [comment]
PR should be feature complete now :)

**Review by MrChico — APPROVED** [review]

**Comment by MrChico** [comment]
Looks good but needs rebase!

**Comment by TransmissionsDev** [comment]
> Looks good but needs rebase!

gotchu

**Review by MrChico — APPROVED** [review]

**Comment by MrChico** [comment]
Thank you!

**Pull request merged**